plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -299,22 +342,60 @@ But in our case this is okay, because:
 3. We *do* have ordering guarantees between the throws inside the ST action and the catch, since they are ultimately using the same thread token.
 -}
 
-
--- | Less-polymorphic synonym for 'throwM', useful in this module
-throwCek :: (PrettyUni uni fun) => CekEvaluationException uni fun -> CekM s x
--- See Note [Throwing exceptions in ST]
-throwCek = throwM
-
--- | Less-polymorphic, exception-based version of 'throwing', useful in this module
-throwingCek :: forall uni fun t s x . (PrettyUni uni fun) => AReview (CekEvaluationException uni fun) t -> t -> CekM s x
-throwingCek l = reviews l throwM
-
 -- | Call 'dischargeCekValue' over the received 'CekVal' and feed the resulting 'Term' to
 -- 'throwingWithCause' as the cause of the failure.
 throwingDischarged
     :: (PrettyUni uni fun)
-    => AReview (EvaluationError CekUserError (MachineError fun (Term Name uni fun ()))) t -> t -> CekValue uni fun -> CekM s x
-throwingDischarged l t = throwingWithCauseExc l t . Just . void . dischargeCekValue
+    => AReview (EvaluationError CekUserError (MachineError fun)) t
+    -> t
+    -> CekValue uni fun
+    -> CekM uni fun s x
+throwingDischarged l t = throwingWithCause l t . Just . dischargeCekValue
+
+-- | Enable throwing 'CekValue' rather than 'Term' within the received action.
+withErrorDischarging
+    :: CekCarryingM (CekValue uni fun) uni fun s a
+    -> CekCarryingM (Term Name uni fun ()) uni fun s a
+withErrorDischarging = coerce
+
+-- | Handle a 'CekEvaluationException' in the 'CekCarryingM' monad.
+catchErrorCekCarryingM
+    :: PrettyUni uni fun
+    => CekCarryingM term uni fun s a
+    -> (CekEvaluationException uni fun -> CekCarryingM term uni fun s a)
+    -> CekCarryingM term uni fun s a
+catchErrorCekCarryingM a h =
+    CekCarryingM . unsafeIOToST $ unsafeRunCekCarryingM a `catch` (unsafeRunCekCarryingM . h) where
+        -- | Unsafely run a 'CekCarryingM' computation in the 'IO' monad by converting the
+        -- underlying 'ST' to it.
+        unsafeRunCekCarryingM :: CekCarryingM term uni fun s a -> IO a
+        unsafeRunCekCarryingM = unsafeSTToIO . unCekCarryingM
+
+instance PrettyUni uni fun => MonadError (CekEvaluationException uni fun) (CekM uni fun s) where
+    throwError = CekCarryingM . throwM
+    catchError = catchErrorCekCarryingM
+
+instance PrettyUni uni fun =>
+            MonadError
+                (CekEvaluationExceptionCarrying (CekValue uni fun) fun)
+                (CekCarryingM (CekValue uni fun) uni fun s) where
+    throwError = CekCarryingM . throwM . mapCauseInMachineException dischargeCekValue
```

Comment:
Yep, we always throw terms. I clearly should've documented that, will do.

Reply:
> But then at the top level we could catch both the errors with terms and the errors with values and convert the latter into the former.

That does seem cleaner than what I wrote, I'll implement it.

> We could even keep it cleaner, and teach the builtin machinery to throw a completely different type of error

Not sure what you mean.

Reply:
Ah, thanks, I get it now. Still don't get to to implement that, though. The builtin application machinery gets invoked in the `CekM`-ish monad as well as, well, the CEK machine itself, so it seems that these two do need to share (for the most part) the type of exceptions they throw, if we're going to have a single `MonadError` instance?

Reply:
Yeah I guess.

Reply:
Could run it in a different monad that's coercible to CekM with a different instance :shrug:

Reply:
... and then instead of having two different monads and duplicating `Functor` etc instances we could just have a single monad parameterized by the type errors that it allows us to throw, at which point we arrived at the current solution with `CekCarryingM` (modulo the one/two `MonadError` instances choice) :)